src/server/src/main/java/io/cassandrareaper/storage/memory/ReplicaLockManagerWithTtl.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2024-2024 DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.cassandrareaper.storage.memory;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import com.google.common.annotations.VisibleForTesting;
+
+public class ReplicaLockManagerWithTtl {
+
+  private final ConcurrentHashMap<String, LockInfo> replicaLocks = new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<UUID, Set<UUID>> repairRunToSegmentLocks = new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<UUID, ReentrantLock> runIdLocks = new ConcurrentHashMap<>();
+
+  private final long ttlMilliSeconds;
+
+  public ReplicaLockManagerWithTtl(long ttlMilliSeconds) {
+    this.ttlMilliSeconds = ttlMilliSeconds;
+    // Schedule cleanup of expired locks
+    ScheduledExecutorService lockCleanupScheduler = Executors.newScheduledThreadPool(1);
+    lockCleanupScheduler.scheduleAtFixedRate(this::cleanupExpiredLocks, 1, 1, TimeUnit.SECONDS);
+  }
+
+  private String getReplicaLockKey(String replica, UUID runId) {
+    return replica + runId;
+  }
+
+  private Lock getLockForRunId(UUID runId) {
+    return runIdLocks.computeIfAbsent(runId, k -> new ReentrantLock());
+  }
+
+  public boolean lockRunningRepairsForNodes(UUID runId, UUID segmentId, Set<String> replicas) {
+    Lock lock = getLockForRunId(runId);
+    lock.lock();
+    try {
+      long currentTime = System.currentTimeMillis();
+      // Check if any replica is already locked by another runId
```

Comment:
Issue: This is confusing. I think this is checking if each replica is locked and then taking action per replica. It isn't checking if ANY replica is locked.

Reply:
I don't follow your reasoning. We won't be able to run the segment if "any" of the replicas is already locked. Not if each of them is locked. That would mean we would be allowed to run if some of them are locked and no others, that's not the case.

Reply:
Right, I think this came about from me misunderstanding what you were trying to do here. You basically want to check if a given replica for a specific segment is locked, not if any replica for any segment is locked.